---
title: "ChromeQC: Library Quality of Linked Reads"
output:
  flexdashboard::flex_dashboard:
    theme: bootstrap
---

```{r setup, include=FALSE}
library(dplyr)
library(flexdashboard)
library(ggplot2)
library(here)
library(knitr)
library(readr)
library(scales)
library(tibble)

knitr::opts_chunk$set(echo = FALSE)
```

# Summary

## Column

```{r read-molecules, include=F}
molecules_orig <-
    read_tsv(
        here("hg004g1.lrbasic.bwa.sortbx.molecule.tsv"),
		col_types = cols(
		  Rname = col_character(),
		  Start = col_integer(),
		  End = col_integer(),
		  BX = col_character(),
		  Reads = col_integer(),
		  MI = col_integer()
		)
	) %>%
	mutate(
		Pairs = as.integer(round(Reads / 2)),
		Observed_size = End - Start) %>%
	filter(Reads >= 4, Observed_size >= 500) %>%
	mutate(Size = round(Observed_size * (1 + 2 / (Pairs - 1)) - 1))

molecules <- molecules_orig %>%
	filter(Reads >= 6)
```

### Input DNA
```{r total-dna, echo=F}
ggplot(molecules) +
	aes(x = Size, weight = Size) +
	geom_histogram(binwidth = 5e3, boundary = 0) +
	scale_x_continuous(name = "Molecule size",
		labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Total DNA mass",
		labels = unit_format(unit = "Mbp", scale = 1e-6)) +
	coord_cartesian(x = c(0, 250e3)) +
	theme_bw(18)
```

### Molecule size distribution
```{r molecule-size, echo=F}
Nx <- function(x, p) {
	sorted <- rev(sort(x))
	sorted[cumsum(sorted) >= sum(sorted) * p][1]
}

N50 <- function(x) Nx(x, p = 0.5)

molecule_stats <- function(x)
	summarize(x,
		Molecules = n(),
		Barcodes = n_distinct(BX),
		N50_LPM = N50(Reads / 2),
		Weighted_mean = weighted.mean(Size, Size),
		N75 = Nx(Size, 0.75),
		N50 = Nx(Size, 0.50),
		N25 = Nx(Size, 0.25),
		Max = max(Size))

unfiltered_stats <- molecule_stats(molecules_orig)
filtered_stats <- molecule_stats(molecules)

kable(rbind(
	unfiltered_stats %>% add_column(Filter = "Reads ≥ 4", .before = 1),
	filtered_stats %>% add_column(Filter = "Reads ≥ 6", .before = 1)))
```

Long Ranger
===================================== 

```{r prelims}
summary_csv <- here("hg004g1_lrwgsvc/outs/summary.csv")
g2 <- read.csv(summary_csv)

# Format a number
digi <- function(num){
  format(num, digits = 2, big.mark=",")
}
```

<!--The goal was to display the quants and statistics from Long Ranger Summary file in a user-friendly manner.
Metrics were classified into relevant category pertaining to molecule or alignment.
All numbers were formatted so that
-Large numbers had comma separators (digi function)
-Decimals did not exceed two significant digits (digi function)
-Fractions were converted to percentages
-Large values in bases or base pairs were converted to kb or kbp
Plots don't make sense for these values; can think of value boxes 
or guages for important percentage values-->

Sidebar {.sidebar}
-------------------------------------

File: `r summary_csv` 

***

Chromium `r g2$instrument_ids`

***

Long Ranger v. `r g2$longranger_version`

Column {data-width=7000}
-------------------------------------

**GEMs**  

*Number of GEMs* \ \t \ \t\ \t \ \t     `r digi(g2$gems_detected)`  


*Mean DNA per GEM* \ \t \ \t\ \t \ \t  `r digi(g2$mean_dna_per_gem)` read pairs

**INPUT DNA**

*DNA per well* \ \t \ \t\ \t \ \t `r digi(g2$corrected_loaded_mass_ng)` ng  

*Mean length +/- s.d.* \ \t \ \t\ \t \ \t  `r digi(g2$molecule_length_mean/1000)` kbp +/- `r digi(g2$molecule_length_stddev/1000)` kbp

**BARCODE (BC)** 

*Reads with valid BC* \ \t \ \t\ \t \ \t `r digi(g2$bc_on_whitelist*100)` %

*Mean base quality score, BC bases*  \ \t \ \t\ \t \ \t`r digi(g2$bc_mean_qscore)`

*N50, read-pairs per input DNA molecule* \ \t \ \t\ \t \ \t `r g2$n50_linked_reads_per_molecule`


Column {data-width=7000}
-------------------------------------

**STRUCTURAL VARIANT CALLS**

*Large SVs* \ \t \ \t\ \t \ \t `r digi(g2$large_sv_calls)`  

*Deletions* \ \t \ \t\ \t \ \t `r digi(g2$short_deletion_calls)`

**PHASING** 

*Phased SNPs* \ \t \ \t\ \t \ \t `r digi(g2$snps_phased*100)` %

*Genes phased in single phase block* \ \t \ \t\ \t \ \t `r digi(g2$genes_phased_lt_100kb*100)` %

*Longest phase block* \ \t \ \t\ \t \ \t `r digi(g2$longest_phase_block/1000)` kb

*N50 length, called phased blocks* \ \t \ \t\ \t \ \t `r digi(g2$n50_phase_block/1000)` kb

**LINKED-READ ALIGNMENT** 

*Long Ranger input reads* \ \t \ \t\ \t \ \t `r digi(g2$number_reads)`

*Mapped reads* \ \t \ \t\ \t \ \t `r digi(g2$mapped_reads)`

*Median insert size of aligned read pairs* \ \t \ \t\ \t \ \t `r digi(g2$median_insert_size)`

*Mean read depth* \ \t \ \t\ \t \ \t `r digi(g2$mean_depth)`

*Reads marked as PCR duplicates* \ \t \ \t\ \t \ \t `r digi(g2$pcr_duplication*100)` %  

*Non-N bases with zero coverage* \ \t \ \t\ \t \ \t `r digi(g2$zero_coverage*100) ` %   

*Aligned bases mapped to target regions* \ \t \ \t\ \t \ \t `r digi(g2$on_target_bases*100)` % 

Sequencing Quality
===================================== 

Column
-------------------------------------

### Read 1 Q20

```{r}
gauge(digi(g2$r1_q20_bases_fract*100), min = 0, max = 100, symbol = '%')
```

### Read 1 Q30

```{r}
gauge(digi(g2$r1_q30_bases_fract*100), min = 0, max = 100, symbol = '%')
```

### Read 2 Q20

```{r}
gauge(digi(g2$r2_q20_bases_fract*100), min = 0, max = 100, symbol = '%')
```

### Read 2 Q30

```{r}
gauge(digi(g2$r2_q30_bases_fract*100), min = 0, max = 100, symbol = '%')
```

Column
-------------------------------------

### Sample Index Q20

```{r}
gauge(digi(g2$si_q20_bases_fract*100), min = 0, max = 100, symbol = '%')
```

### Sample Index Q30

```{r}
gauge(digi(g2$si_q30_bases_fract*100), min = 0, max = 100, symbol = '%')
```

### Barcode Q20

```{r}
gauge(digi(g2$bc_q20_bases_fract*100), min = 0, max = 100, symbol = '%')
```

### Barcode Q30

```{r}
gauge(digi(g2$bc_q30_bases_fract*100), min = 0, max = 100, symbol = '%')
```

```{r read alignmentstats}
molecule_tsv <- here("hg004g1.lrbasic.bwa.sortbx.molecule.tsv")
molecules_orig <- read.table(molecule_tsv, sep="\t", header = TRUE)
```

Barcodes
===================================== 

Sidebar {.sidebar}
-------------------------------------
File: 

***

Chromium `r g2$instrument_ids`

***

Alignment Program Name

```{r funchist}
#Function for plotting with ggplot
makehist <- function(obj, varname, weight, binwidth, fill, color, ti, xname, yname, int) {
ggplot(obj,aes(x=varname, weight=weight)) + geom_histogram(binwidth = binwidth, fill=fill, color=color) + ggtitle(ti) +
scale_x_continuous(name=xname) + scale_y_continuous(name =yname) + geom_vline(xintercept = int, colour="red", linetype = "dashed") + theme_minimal()
}
```

### Read Distribution of Unmatched Barcodes

```{r unmatched}
ubc_tsv <- here("select_random_subset/unmatched_barcodes_population_counts.tsv")
ubc <- read_tsv(ubc_tsv)
colnames(ubc) <- c("Read_count", "Barcode_count")
tableubc <- ubc %>% group_by(Barcode_count) %>% summarize(sum(Read_count))
colnames(tableubc) <- c("Barcode_count", "Read_count")
knitr::kable(tableubc)
```

### Read Distribution of White-listed Barcodes

```{r whitelist}
wbc_tsv <- here("select_random_subset/whitelist_barcodes_population_counts.tsv")
wbc <- read_tsv(wbc_tsv)
colnames(wbc) <- c("Read_count", "Barcode_count")
tablewbc <- wbc %>% group_by(Barcode_count) %>% summarize(sum(Read_count))
colnames(tablewbc) <- c("Barcode_count", "Read_count")
knitr::kable(tablewbc, big.mark=",")
```

Row {data-height=200}
-------------------------------------

```{r}
ubc$Barcode_count <- factor(ubc$Barcode_count)
ggplot(ubc, aes(x=Barcode_count, y=Read_count)) + geom_bar(stat="identity")
```

```{r}
wbc$Barcode_count <- factor(wbc$Barcode_count)
ggplot(wbc, aes(x=Barcode_count, y=Read_count)) + geom_bar(stat="identity") 
```

Molecules
===================================== 

Sidebar {.sidebar}
-------------------------------------
File: 

***

Chromium `r g2$instrument_ids`

***

Alignment Program Name

```{r plotting}
#introduce a column with molecule size
molecules <- molecules_orig %>% mutate(Size = End - Start)

# function for calculating bin size?
#breaks <- pretty(range(x), n = nclass.FD(x), min.n = 1)
#bwidth <- breaks[2]-breaks[1]
```

Calculations from SAM files

Row {.tabset .tabset-fade, data-height=200}
-------------------------------------

### Size of the molecules 

```{r}
makehist(molecules, molecules$Size, molecules$Size, 5000,"orange", "green", "Size Distribution of Molecules", "Molecule size in bps", "Number of molecules", mean(molecules$Size))
```

### Number of reads per molecule

```{r}
makehist(molecules, molecules$Reads, 1, 2,"blue", "green", "Read counts per molecule", "Number of Molecules", "Number of Reads", mean(molecules$Reads))
```

Column {.tabset .tabset-fade, data-height=200}
-------------------------------------

#Number of reads per barcode 

```{r}
BClist <- molecules %>% group_by(BX) %>% summarize(counts=sum(Reads))
#sum(BClist$counts)
#sum(molecules$Reads)
makehist(BClist, BClist$counts, 1, 5,"light blue", "green", "Number of reads per barcode", "Number of Barcodes", "Number of reads", mean(BClist$counts))
```

#Number of molecules per barcode 

```{r}
molBC <- molecules %>% group_by(BX) %>% summarize(molecules=length(MI))
makehist(molBC, molBC$molecules, 1, 1,"brown", "green", "Number of molecules per barcode", "Number of Barcodes", "Number of molecules", mean(molBC$molecules))
```

<!--TO DO 
Function for histogram bin size?
Proofing for technical accuracy
Map the metric terms used on page 1 to Long Ranger summary metrics-may be provide a link to https://support.10xgenomics.com/genome-exome/software/pipelines/latest/output/metrics for further explanation
or mouse over?? -->
